### PR TITLE
WIP: fix stat.isWritable() regression on Windows

### PR DESCRIPTION
**Pull request by perlun**
Note: This is clearly not ready for merge, consider it WIP and a candidate for collaboration.

I've been looking at this issue quite extensively tonight. First fixed one simple test that was failing on Windows 8 and Windows 10 (since the comparison wouldn't include these OS:es; I simplified that part since the previous approach makes it unnecessarily hard to support different Windows versions).

Then, coming to the actual propblem (https://github.com/jruby/jruby/issues/3505) - the weird thing is that the `File.stat('foo').writable?` call in e.g. [File.tmpdir](https://github.com/jruby/jruby/blob/jruby-1_7/lib/ruby/shared/tmpdir.rb#L33) fails. But when I try to reproduce the problem in the `jnr` repo, all is well and things work far too well...

I guess I would have to point my JRuby (using `jruby-complete` unfortunately, which makes it a bit harder) to the `jnr-posix.jar` file I am compiling, to ensure that my JRuby runs towards the same code as the FileStatTest...

Any other obvious suggestions on how to proceed with this?

### CODE REVIEW

File: src/test/java/jnr/posix/FileStatTest.java [hunk 1]
```diff
@@ -157,4 +155,12 @@ public void filestatFailed() throws Throwable {
             }
         }
     }
+
+    @Test public void statIsWritable() throws Throwable {
+      String tempDirectory = System.getenv("TEMP");
+      FileStat st = posix.stat(tempDirectory);
+
+      // FIXME: This fails on JRuby, but succeeds here. Why?
```
Comment (nirvdrum) [inline src/test/java/jnr/posix/FileStatTest.java]: Having not really looked at this at all, Ruby applies extra checks around the permissions of the `$ENV['TMPDIR']` directory to prevent security issues. One is checking that the directory isn't world writable, I believe.

Comment (perlun) [inline src/test/java/jnr/posix/FileStatTest.java]: Do you happen to have any idea of where that check would be? When trying to understand (the somewhat complex) JRuby part of this, I got the impression that it pretty much just forwards the call to `FileStat.isWritable()`: https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/RubyFileStat.java#L493

Comment (perlun) [inline src/test/java/jnr/posix/FileStatTest.java]: Also, very simple calls like this fails:

```
irb(main):001:0> File.stat('c:/temp').writable?
=> false
```

Comment (enebo) [inline src/test/java/jnr/posix/FileStatTest.java]: Are you sure win8 and win10 have different values?  I am not sure if this platform work is needed or not?

I did not come up with a fix but I thought I confirmed that CommonFileInformation.getMode() was returning the wrong number.  Seems like perhaps something with this modulo logic is wrong or something.  If you look at MRI source code it feels like I followed that logic but something is not right. 

To recap, I think this is not platform specific at all and I think the mode calculated is not right.  The mode is only done in the getMode() method mentioned above and also in WindowsRawFileStat.setup.  Looking at how MRI calculates this will be helpful because I looked to them for guidance on getting these unix mode values right...

Comment (perlun) [inline src/test/java/jnr/posix/FileStatTest.java]: > Are you sure win8 and win10 have different values? I am not sure if this platform work is needed or not?

Yes, they do in fact. I ran it in a Win8.1 and a Win10 VM to test that part. It behaves differently w/ different JVM versions though.

Thanks for the suggestion. **But**, the weird thing is that the test below succeeds - i.e. when running `jnr-posix` in itself, the folder is considered writable. When jruby calls this code, it fails.

Comment (nirvdrum) [inline src/test/java/jnr/posix/FileStatTest.java]: Having dug up the previous issue, I don't think this is related. But here's an example of Ruby applying additional logic:

https://github.com/jruby/jruby/blob/a968fc7db36d49ca3e09201f788bce7285d7231e/lib/ruby/stdlib/tmpdir.rb#L30-L38

**Comment by headius** [comment]
Pinging here as well for updates on status.

**Comment by enebo** [comment]
This is caused by the jnr issue since JRuby itself does not do anything other than use the mode value coming from jnr-posix.

**Comment by perlun** [comment]
Yep. @enebo - how have you been doing when developing jnr-posix and wishing to test your JRuby towards a specific (local) version of that code? Is it just a matter of `mvn install` in the one repo and then `mvn build` in the other?

**Comment by enebo** [comment]
@perlun It is so long as you update core/pom.rb in jruby to indicate that you want the same version (which will be 3.0.24-SNAPSHOT) and then run ./mvnw once to regenerate the pom.xml.  once things are updated you then just run 'mvn' (not 'mvn build') and you should be good.

**Comment by perlun** [comment]
This PR has now gone stale. The `writable` bug was indeed not a bug in jnr-posix, but just a "forgetting to update jnr-posix reference" as you know.

The other parts (detect Windows 8 and 10 correctly) I think we should keep, because some of the tests currently fail on Windows and test failures are never good. Will sort that out in a while, sometime later this year (it's not urgent).

**Comment by enebo** [comment]
@perlun thanks for looking into it even though it ended up being my silly mistake.  Hopefully this will help for the next problem we have in jnr-posix :)
